frame/pallet-xvm/src/lib.rs

```diff
@@ -0,0 +1,174 @@
+//! # XVM pallet
+//!
+//! ## Overview
+//!
+//!
+//! ## Interface
+//!
+//! ### Dispatchable Function
+//!
+//!
+//! ### Other
+//!
+//!
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use codec::{Decode, Encode};
+use sp_runtime::{traits::Member, RuntimeDebug};
+use sp_std::prelude::*;
+
+pub mod pallet;
+pub use pallet::pallet::*;
+
+/// EVM call adapter instance.
+#[cfg(feature = "evm")]
+pub mod evm;
+
+/// Wasm call adapter instance.
+#[cfg(feature = "wasm")]
+pub mod wasm;
+
+/// XVM context consist of unique ID and optional execution arguments.
+#[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug, scale_info::TypeInfo)]
+pub struct XvmContext<VmId> {
+    /// VM identifier.
+    pub id: VmId,
+    /// Encoded VM execution environment.
+    pub env: Option<Vec<u8>>,
+}
+
+/// The engine that support synchronous smart contract execution.
+/// For example, EVM.
+///
+/// An impl code should realize input data conversion using provided
+/// metadata.
+pub trait SyncVM<VmId, AccountId> {
+    /// Unique VM identifier.
+    fn id() -> VmId;
+
+    /// Make a call to VM contract and return result or error.
+    ///
+    ///
+    fn xvm_call(
+        context: XvmContext<VmId>,
+        from: AccountId,
+        to: Vec<u8>,
```

Comment:
What is `to`? Does it combine address and function identifier?

Reply:
Just destination ID, may be address, may be MultiAddress, may be MultiLocation

Reply:
Do you mean that depending on the `VmId`, smart contract will have to encode destination differently?
No common type which XVM converts to appropriate type for the destination VM?

Reply:
Yes, depend of VM destination could be different. May be something like multi-location is more concrete but generic way of destination encoding. Not sure I want bring XCM only for this primitive.

Reply:
It’s open question anyway, not decided yet how better to present it.